### PR TITLE
Event pattern. Fixes #243

### LINKED ISSUE #243: Events: Discuss the Event Pattern
This is a sub-task for #64 

Write the article that discusses the event pattern in .NET

### PR DESCRIPTION
**Pull request by BillWagner**
Fixes #243 
This relies on PR #246, which creates the placeholders.

### CODE REVIEW

File: docs/languages/csharp/event-pattern.md [hunk 1]
```diff
@@ -1,9 +1,344 @@
 # The Standard .NET Event Pattern
 
+By [Bill Wagner](https://github.com/BillWagner)
+
 [Previous](events-overview.md)
 
-This article will the standard .NET event pattern.
+.NET events generally follow a few known patterns. Standardizing
+on these patterns means that developers can leverage knowledge of
+those standard patterns can be applied to any .NET event program.
```
Comment (ghogen) [inline docs/languages/csharp/event-pattern.md]: standard patterns, which can be applied...

File: docs/languages/csharp/event-pattern.md [hunk 2]
```diff
@@ -1,9 +1,344 @@
...
+void Handler(object sender, EventArgs args);
+```
+
+The return type is void. Events are based on delegates and are
+multicast delegates. That supports multiple subscribers for any
+event source. The single return value from a method doesn't scale
+to multiple event subscribers. Which return value does the event
+source see after raising an event? Later in this article you'll
+see how to create event protocols that support event subscribers
+reporting information to the event source.
```
Comment (ghogen) [inline docs/languages/csharp/event-pattern.md]: reporting -> that report (suggested change)

File: docs/languages/csharp/event-pattern.md [hunk 3]
```diff
@@ -1,9 +1,344 @@
...
+The second argument has typically been a type that is derived from
+`System.EventArgs`. (You'll see in the 
+[next section](modern-events.md) that this convention is no longer
+enforced.) If your event type does not need any additional
+arguments, you will still provide both arguments.
+There is a special value, `EventArgs.Empty` that you should use to
+denote that your event does not contain any additional information.
+
+Let's build a class that lists files in a directory, or any
+subdirectory that follow a pattern. This component raises an event
```
Comment (ghogen) [inline docs/languages/csharp/event-pattern.md]: suggested revision: files in a directory or any of its subdirectories that follow a pattern

File: docs/languages/csharp/event-pattern.md [hunk 4]
```diff
@@ -1,9 +1,344 @@
...
+    public string FoundFile { get; }
+
+    public FileFoundArgs(string fileName)
+    {
+        FoundFile = fileName;
+    }
+}
+```
+
+Even though this type looks like a small, data only type, you should
```
Comment (ghogen) [inline docs/languages/csharp/event-pattern.md]: hyphenate data-only

File: docs/languages/csharp/event-pattern.md [hunk 5]
```diff
@@ -1,9 +1,344 @@
...
+    {
+        foreach (var file in Directory.EnumerateFiles(directory, searchPattern))
+        {
+            OnFoundFile?.Invoke(this, new FileFoundArgs(file));
+        }
+    }
+}
+```
+
+# Definining and Raising Field Like Events
```
Comment (ghogen) [inline docs/languages/csharp/event-pattern.md]: Field-like hyphenate

File: docs/languages/csharp/event-pattern.md [hunk 6]
```diff
@@ -1,9 +1,344 @@
...
+
+The simplest way to add an event to your class is to declare that
+event as a public field, as in the above example:
+
+```cs
+public event EventHandler<FileFoundArgs> OnFoundFile;
+```
+
+This looks like it's declaring a public field, which would appear to
+be bad Object Oriented practice. You want to protect data access
```
Comment (ghogen) [inline docs/languages/csharp/event-pattern.md]: bad object-oriented practice

File: docs/languages/csharp/event-pattern.md [hunk 7]
```diff
@@ -1,9 +1,344 @@
...
+through properties, or methods. While this make look like a bad
+practice, the code generated by the compiler does create wrappers so
+that the event objects can only be accessed in safe ways. The only
+operations available on a Field like event are add handler:
```
Comment (ghogen) [inline docs/languages/csharp/event-pattern.md]: field-like (lowercase and hyphenated)
set off the operation "add handler" in some way, maybe in double quotes to make it clear that it's an operation

**Comment by BillWagner** [comment]
Responded to edits, rebased and squashed.

**Pull request merged**